### PR TITLE
[3.4][Feature][Ready] Select2_from_ajax can depend on any other input

### PR DESCRIPTION
**Pull request by tabacitu**
This adds a parameter to the AJAX in ```select2_from_ajax``` and ```select2_from_ajax_multiple```: this parameter is called ```form``` and it includes all form inputs at the time the AJAX call is made. I don't like to toot my own horn... (yes I do)... but I think this is a _pretty brilliant_ way of achieving select2s that are dependant on something else :-) 

**Reasoning:** To filter the options in a ```select2_from_ajax``` depending on another input, the only thing we need in the results method in PHP is _all form inputs_. If we have those, the world is our oyster. We can filter the results of a ```select2_from_ajax``` depending on a select, on a checkbox, on a text input, or _multiple inputs_ together.

This is a non-breaking change. So no change to syntax, or anything. The only place you need to edit to make a ```select2_from_ajax``` dependant on another form input is the PHP method where you return the results.

Here's an example of a ```select2_from_ajax``` field that depends on a ```category``` select2 above:
```php
    public function index(Request $request)
    {
        $search_term = $request->input('q');
        $form = collect($request->input('form'))->pluck('value', 'name');

        $options = Article::query();

        // if no category has been selected, show no options
        if (! $form['category']) {
            return [];
        }

        // if a category has been selected, only show articles in that category
        if ($form['category']) {
            $options = $options->where('category_id', $form['select2']);
        }

        // if a search term has been given, filter results to match the search term
        if ($search_term) {
            $options = $options->where('title', 'LIKE', '%'.$search_term.'%');
        }

        return $options->paginate(10);
    }
```

@OwenMelbz I know you've wanted this for a while now. What do you think? 
@Laravel-Backpack/veterans - any possible side-effects I'm not seeing?

Pretty excited about this one :-)

### CODE REVIEW

**Comment by OliverZiegler** [comment]
@tabacitu what about changing dependencies after selection? Will this somehow trigger an Ajax reload?

Say:

You have a  locality state select and a new Ajax city select.
I e.g select Germany as state and afterwards the Ajax only shows german cities. Now I select Munich.

Now I changed my opinion and select France as state. Will the city be reset? Or can I submit my form with France - Munich? (Which could be checked for validity after submit, but this is kind of inconvenient)

**Comment by tabacitu** [comment]
Hmm... you're right. The current change would NOT unselect the second field... and intuitively it should.... hmm... good point, thank you @OliverZiegler ... let me think about it...

**Comment by tabacitu** [comment]
Note to self: 

### Option 1 - in the AJAX method

When returning the results in AJAX, we can do something like:

```diff
-        return $options->paginate(10);

+        $results = $options->paginate(10)->toArray();
+        $results['clear_inputs'] = ['something', 'something_else'];
+
+        return $results;
```

Then we'd know which inputs to reset.

### Option 2 - in the field definition

```diff
        $this->crud->addField([ // select2_from_ajax: 1-n relationship
            'label'                => "Country", // Table column heading
            'type'                 => 'select2_from_ajax',
            'name'                 => 'select2_from_ajax', // the column that contains the ID of that connected entity;
            'entity'               => 'country', // the method that defines the relationship in your Model
            'attribute'            => 'name', // foreign key attribute that is shown to user
            'model'                => "App\Models\Country", // foreign key model
            'data_source'          => url('api/country'), // url to controller search function (with /{id} should return model)
            'placeholder'          => 'Select a country', // placeholder for the select
            'minimum_input_length' => 2, // minimum characters to type before querying results
+           'inputs_to_reset' => ['city'], // reset these inputs when the selected value changes
        ]);
```

Come to think of it - no contest. Definitely OPTION 2. Both solutions, though, have a problem:
- since you don't know what kind of field that is ```$('input[name=xxx]').val("")``` will probably not work for a few input types; ok, we can find ways around for checkbox, textarea, etc, but it probably won't work for more complicated field types (table, etc);

Food for thought. 

Sorry for spamming on this, I'm just sharing my train of thoughts because today I can only work in small increments and I don't want to lose progress.

**Comment by imansugirman** [comment]
Hi @tabacitu  I Follow u'r instruction like this

https://gist.github.com/imansugirman/07268545cd93e4da61e7d59ecd278c0a

but show error access api /api/citysearch

jepretan layar 2018-07-03 pada 03 02 05

**Comment by tabacitu** [comment]
@imansugirman for the moment please use ```        return $options->paginate(10);``` - ignore my suggestion in the last comment, that's not implemented yet.

**Comment by elricho** [comment]
Hi @tabacitu - Food for thought : On large forms the "form: form.serializeArray()" inside the Ajax call can blow out the URL length resulting in "414 URI Too Long" errors. 

As a workaround we added a  'watched_field' key to the $fields array. The watched_field then becomes a trigger for change events inside the javascript & replaces "form: form.serializeArray()" with "{{ $field['watched_field'] }}: $('[name={{ $field['watched_field'] }}]').val()" in the Ajax call. 

Hope that helps. :-)

**Comment by tva3777** [comment]
is there a final solution for the dependent selects anyway?

**Comment by markcollister** [comment]
Just FYI, I like this dependant select2 implementation. 

I came across a kinda edge case where I had an image field type in the form which was being serialized

**Comment by GonrasK** [comment]
I published and overwrote the select2_from_ajax using `php artisan backpack:crud:publish fields/select2_from_ajax` and still I'm not sure where am i supposed to push this change mentioned above. I have a campaign crud controller which looks something like this - 

`  $this->crud->addField([
		    'name' => 'industry_id',
		    'entity' => 'industry',
		    'type' => 'select2',
		    'attribute' => 'name',
		    'label' => "Industry",
	    ]);
	    $this->crud->addField([
		    'name' => 'country_id',
		    'entity' => 'country',
		    'type' => 'select2',
		    'attribute' => 'country_name',
		    'label' => "Country",
	    ]);

	    $this->crud->addField([
		    'name' => 'website_id',
		    'entity' => 'website',
		    'type' => 'select2',
		    'attribute' => 'name',
		    'label' => "Website",
	    ]);
	    $this->crud->addField([
		    'name' => 'campaign_type_id',
		    'entity' => 'campaign_type',
		    'type' => 'select2',
		    'attribute' => 'name',
		    'label' => "Campaign Type",
	    ]);

	    $this->crud->addField([
		    'name' => 'site_page_id',
		    'entity' => 'site_page',
		    'type' => 'select2',
		    'attribute' => 'name',
		    'label' => "Site Page",
	    ]);`

Where the 'website' entity is depended on the industry field, and the 'site_page_id' is depended on the cross between the 'website' entity and the 'campaign_type' entity.

**Comment by tabacitu** [comment]
@GonrasK to use the changes in this PR before it gets merged:
- publish the select2_from_ajax blade file with ```php artisan backpack:crud:publish fields/select2_from_ajax``` (I see you've already done that)
- open that file (```resources/views/vendor/backpack/crud/fields/select2_from_ajax.blade.php```) and overwrite its contents with the final version as this PR changes it (click [here](https://github.com/Laravel-Backpack/CRUD/blob/fca7cfe16b9128579127316f8afd9998261e3fc7/src/resources/views/fields/select2_from_ajax.blade.php) or when seeing the differences click View);

That's it.

**Comment by tabacitu** [comment]
Just pushed the update. I ended up calling it “_dependencies_”. Final instructions on how to use the field below:

——

Say you want to show two selects:
- the first one shows Categories
- the second one shows Articles, but only from the category above

1. In you CrudController you would do:

```php

        $this->crud->addField([    // SELECT2
            'label'         => ‘Category',
            'type'          => 'select',
            'name'          => ‘category',
            'entity'        => 'category',
            'attribute'     => 'name',
        ]);

        $this->crud->addField([ // select2_from_ajax: 1-n relationship
            'label'                => "Article", // Table column heading
            'type'                 => 'select2_from_ajax_multiple',
            'name'                 => ‘articles', // the column that contains the ID of that connected entity;
            'entity'               => 'article', // the method that defines the relationship in your Model
            'attribute'            => 'title', // foreign key attribute that is shown to user
            'data_source'          => url('api/article'), // url to controller search function (with /{id} should return model)
            'placeholder'          => 'Select an article', // placeholder for the select
            'minimum_input_length' => 0, // minimum characters to type before querying results
            'dependencies'         => [‘category’], // when a dependency changes, this select2 is reset to null
            // ‘method'                    => ‘GET’, // optional - HTTP method to use for the AJAX call (GET, POST)
        ]);
```

That second select points to routes that need to be registered:

```php
Route::get('api/article', 'App\Http\Controllers\Api\ArticleController@index');
Route::get('api/article/{id}', 'App\Http\Controllers\Api\ArticleController@show’);
```

Then that controller would look like this:
```php
<?php

namespace App\Http\Controllers\Api;

use App\Http\Controllers\Controller;
use Backpack\NewsCRUD\app\Models\Article;
use Illuminate\Http\Request;

class ArticleController extends Controller
{
    public function index(Request $request)
    {
        $search_term = $request->input('q');
        $form = collect($request->input('form'))->pluck('value', 'name');

        $options = Article::query();

        // if no category has been selected, show no options
        if (! $form[‘category']) {
            return [];
        }

        // if a category has been selected, only show articles in that category
        if ($form[‘category']) {
            $options = $options->where('category_id', $form[‘category']);
        }

        if ($search_term) {
            $results = $options->where('title', 'LIKE', '%'.$search_term.'%')->paginate(10);
        } else {
            $results = $options->paginate(10);
        }

        return $options->paginate(10);
    }

    public function show($id)
    {
        return Article::find($id);
    }
}
```

**Comment by tabacitu** [comment]
Mergin into  the ```3.5``` branch. 

**Todo:**
- [ ] add instructions above to documentation
- [ ] add ```method``` attribute to documentation
- [ ] add ```dependencies``` attribute to documentation

**Comment by trip-somers** [comment]
This is super useful when populating a form. Is there a similar option for filters? I can't seem to find it.

**Comment by kamilkozak** [comment]
> This is super useful when populating a form. Is there a similar option for filters? I can't seem to find it.

I also think that dependent filters would be useful.

**Comment by helesjunior** [comment]
Good afternoon @tabacitu ,

How do I use this same structure in Backpack 6.x? I have a table of States and another of Cities and I need the city combo to depend on the state information.

On Controller:  
```
[
     'name' => 'state_id',
     'label' => 'Estado',
    'type' => 'select2',
    'entity' => 'state', // the method that defines the relationship in your Model
    'model' => "App\Models\State", // foreign key model
    'attribute' => 'uf_name', // foreign key attribute that is shown to user
    'allows_null' => true,
    'placeholder' => "Selecione o Estado",
    'options' => (function ($query) {
          return $query->orderBy('uf', 'ASC')->get();
    }),
],
[ // select2_from_ajax: 1-n relationship
    'name' => 'city_id',
    'label' => 'Município', // Table column heading
    'type' => 'select2_from_ajax',
    'ajax' => false,
    'model' => "App\Models\City", // foreign key model
    'entity' => 'city', // the method that defines the relationship in your Model
    'attribute' => 'name', // foreign key attribute that is shown to user
    'data_source' => url('api/city'), // url to controller search function (with /{id} should return model)
    'method' => 'GET', // route method, either GET or POST
    'allows_null' => true,
    'placeholder' => 'Selecione o município', // placeholder for the select
    'minimum_input_length' => 0, // minimum characters to type before querying results
    'dependencies' => ['state_id'],
],
```

On City Api:  
```
    public function index(Request $request)
    {
        $search_term = $request->input('q');
        $form = collect($request->input('form'))->pluck('value', 'name');

        $options = City::query();

        // if no category has been selected, show no options
        if (!$form['state_id']) {
            return [];
        }

        // if a category has been selected, only show articles in that category
        if ($form['state_id']) {
            $options = $options->where('state_id', $form['state_id']);
        }

        if ($search_term) {
            return $options->where('name', 'ilike', '%' . strtoupper($search_term) . '%')
                ->orderBy('name')
                ->paginate(10);
        }

        return $options->paginate(10);
    }
    public function show($id)
    {
        return City::find($id);
    }
```

Return Error:  Undefined array key "state_id"

image


Could you help me with this problem or give me a cool alternative?

**Comment by pxpm** [comment]
Hey @helesjunior I created an issue for your problem https://github.com/Laravel-Backpack/community-forum/discussions/1012

Please follow any updates there, by rule of thumb, **always open a new thread** for your issue. 

Cheers

**Comment by quiquegarcia** [comment]
Hello! somebody knows if this is applicable to repeatable fields? I'm in backpack pro v2.2.29
